bootstrap/utils.go

```diff
@@ -0,0 +1,378 @@
+			break
+		}
+
+		return nil, fmt.Errorf("failed to parse validator key - valid PKCS#1, PKCS#8 ...? %v, %v, %v, %v, %v", errParse0, errParse1, errParse2, errParse3, errParse4)
+	}
+
+	return validator, nil
+}
+
+func AddSignerWithIDFromFile(fn string, kid string, bs *bootstrap) error {
+	fi, err := os.Lstat(fn)
+	if err != nil {
+		return fmt.Errorf("failed load load signer key: %v", err)
+	}
+
+	mode := fi.Mode()
+	switch {
+	case mode.IsDir():
+		return fmt.Errorf("signer key must be a file")
+	}
+
+	// Load file.
+	signerKid, signer, err := LoadSignerFromFile(fn)
+	if err != nil {
+		return err
+	}
+	if kid == "" {
+		kid = signerKid
+	}
+	if kid == "" {
+		// Get ID from file, following symbolic link.
+		var real string
+		if mode&os.ModeSymlink != 0 {
+			real, err = os.Readlink(fn)
+			if err != nil {
+				return err
+			}
+			_, real = filepath.Split(real)
+		} else {
+			real = fi.Name()
+		}
+
+		kid = getKeyIDFromFilename(real)
+	}
+
+	if _, ok := bs.signers[kid]; ok {
+		bs.Cfg.Logger.WithFields(logrus.Fields{
+			"path": fn,
+			"kid":  kid,
+		}).Warnln("skipped as signer with same kid already loaded")
+		return nil
+	} else {
+		bs.Cfg.Logger.WithFields(logrus.Fields{
+			"path": fn,
+			"kid":  kid,
+		}).Debugln("loaded signer key")
+	}
+
+	bs.signers[kid] = signer
+	return nil
+}
+
+func ValidateSigners(bs *bootstrap) error {
+	haveRSA := false
+	haveECDSA := false
+	haveEd25519 := false
+	for _, signer := range bs.signers {
+		switch s := signer.(type) {
+		case *rsa.PrivateKey:
+			// Ensure the private key is not vulnerable with PKCS-1.5 signatures. See
+			// https://paragonie.com/blog/2018/04/protecting-rsa-based-protocols-against-adaptive-chosen-ciphertext-attacks#rsa-anti-bb98
+			// for details.
+			if s.PublicKey.E < 65537 {
+				return fmt.Errorf("RSA signing key with public exponent < 65537")
+			}
+			haveRSA = true
+		case *ecdsa.PrivateKey:
+			haveECDSA = true
+		case ed25519.PrivateKey:
+			haveEd25519 = true
+		default:
+			return fmt.Errorf("unsupported signer type: %v", s)
+		}
+	}
+
+	// Validate signing method
+	switch bs.signingMethod.(type) {
+	case *jwt.SigningMethodRSA:
+		if !haveRSA {
+			return fmt.Errorf("no private key for signing method: %s", bs.signingMethod.Alg())
+		}
+	case *jwt.SigningMethodRSAPSS:
+		if !haveRSA {
+			return fmt.Errorf("no private key for signing method: %s", bs.signingMethod.Alg())
+		}
+	case *jwt.SigningMethodECDSA:
+		if !haveECDSA {
+			return fmt.Errorf("no private key for signing method: %s", bs.signingMethod.Alg())
+		}
+	case *signing.SigningMethodEdwardsCurve:
+		if !haveEd25519 {
+			return fmt.Errorf("no private key for signing method: %s", bs.signingMethod.Alg())
+		}
+	default:
+		return fmt.Errorf("unsupported signing method: %s", bs.signingMethod.Alg())
+	}
+
+	if !haveRSA {
+		bs.Cfg.Logger.Warnln("no RSA signing private key, some clients might not be compatible")
+	}
+
+	return nil
+}
+
+func AddValidatorsFromPath(pn string, bs *bootstrap) error {
```

Comment:
And again - same as above. Please make private. Since this issue happens multiple types, i guess a public `Bootstrap` interface should be added which exposes those exported functions as members.